File: silvamang_api/app/Support/SpeciesTaxonomy.php
<?php

namespace App\Support;

use Illuminate\Support\Str;

final class SpeciesTaxonomy
{
    /**
     * Philippine catalog aliases and known legacy labels.
     *
     * @var array<string, string>
     */
    private const CANONICAL_NAMES = [
        'acanthus ebracteatus' => 'Acanthus ebracteatus',
        'acanthus ilicifolius' => 'Acanthus ilicifolius',
        'acanthus volubilis' => 'Acanthus volubilis',
        'aegiceras corniculatum' => 'Aegiceras corniculatum',
        'aegiceras floridum' => 'Aegiceras floridum',
        'avicennia alba' => 'Avicennia alba',
        'avicennia marina' => 'Avicennia marina',
        'avicennia marina var rumphiana' => 'Avicennia rumphiana',
        'avicennia officinalis' => 'Avicennia officinalis',
        'avicennia rumphiana' => 'Avicennia rumphiana',
        'bruguiera cylindrica' => 'Bruguiera cylindrica',
        'bruguiera gymnorhiza' => 'Bruguiera gymnorrhiza',
        'bruguiera gymnorrhiza' => 'Bruguiera gymnorrhiza',
        'bruguiera sexangola' => 'Bruguiera sexangula',
        'bruguiera sexangula' => 'Bruguiera sexangula',
        'camptostemon philippinense' => 'Camptostemon philippinensis',
        'camptostemon phillipinensis' => 'Camptostemon philippinensis',
        'camptostemon philippinensis' => 'Camptostemon philippinensis',
        'ceriops tagal' => 'Ceriops tagal',
        'ceriops zippeliana' => 'Ceriops zippeliana',
        'excoecaria agallocha' => 'Excoecaria agallocha',
        'heritiera littoralis' => 'Heritiera littoralis',
        'lumnitzera littorea' => 'Lumnitzera littorea',
        'lumnitzera racemosa' => 'Lumnitzera racemosa',
        'nypa fruticans' => 'Nypa fruticans',
        'osbornia octodonta' => 'Osbornia octodonta',
        'pemphis acidula' => 'Pemphis acidula',
        'rhizophora apiculata' => 'Rhizophora apiculata',
        'rhizophora mucronata' => 'Rhizophora mucronata',
        'rhizophora stylosa' => 'Rhizophora stylosa',
        'scyphiphora hydrophyllacea' => 'Scyphiphora hydrophylacea',
        'scyphiphora hydrophylacea' => 'Scyphiphora hydrophylacea',
        'sonneratia alba' => 'Sonneratia alba',
        'sonneratia ovata' => 'Sonneratia ovata',
        // Older Philippine references applied this name to Xylocarpus rumphii.
        'xylocarpus granatum' => 'Xylocarpus granatum',
        'xylocarpus moluccensis' => 'Xylocarpus rumphii',
        'xylocarpus rumphii' => 'Xylocarpus rumphii',
    ];

    public static function canonicalName(mixed $value): string
    {
        $displayName = self::displayName($value);

        return self::CANONICAL_NAMES[self::rawLookupKey($displayName)] ?? $displayName;
    }

    public static function lookupKey(mixed $value): string
    {
        return self::rawLookupKey(self::canonicalName($value));
    }

    private static function displayName(mixed $value): string
    {
        $name = trim(str_replace('_', ' ', (string) $value));

        return preg_replace('/\s+/', ' ', $name) ?? '';
    }

    private static function rawLookupKey(mixed $value): string
    {
        return Str::of(self::displayName($value))
            ->lower()
            ->replaceMatches('/[^a-z0-9]+/', ' ')
            ->squish()
            ->toString();
    }
}

File: silvamang_mobile/lib/shared/utils/species_taxonomy.dart
String canonicalSpeciesName(Object? value) {
  final displayName = value?.toString().replaceAll('_', ' ').trim() ?? '';
  final squished = displayName.replaceAll(RegExp(r'\s+'), ' ');
  final key = _rawSpeciesKey(squished);

  return _canonicalNames[key] ?? squished;
}

String normalizedSpeciesKey(Object? value) {
  return _rawSpeciesKey(canonicalSpeciesName(value));
}

String normalizedSpeciesText(Object? value) {
  var normalized = value
      .toString()
      .toLowerCase()
      .replaceAll('_', ' ')
      .replaceAll(RegExp(r'[^a-z0-9]+'), ' ')
      .trim()
      .replaceAll(RegExp(r'\s+'), ' ');

  final aliases = _canonicalNames.entries.toList()
    ..sort((left, right) => right.key.length.compareTo(left.key.length));

  for (final alias in aliases) {
    final aliasText = alias.key.replaceAll('_', ' ');
    final canonicalText = _rawSpeciesKey(alias.value).replaceAll('_', ' ');
    normalized = normalized.replaceAll(aliasText, canonicalText);
  }

  return normalized;
}

String _rawSpeciesKey(Object? value) {
  return value
      .toString()
      .trim()
      .toLowerCase()
      .replaceAll('_', ' ')
      .replaceAll(RegExp(r'[^a-z0-9]+'), ' ')
      .trim()
      .replaceAll(RegExp(r'\s+'), '_');
}

const _canonicalNames = <String, String>{
  'acanthus_ebracteatus': 'Acanthus ebracteatus',
  'acanthus_ilicifolius': 'Acanthus ilicifolius',
  'acanthus_volubilis': 'Acanthus volubilis',
  'aegiceras_corniculatum': 'Aegiceras corniculatum',
  'aegiceras_floridum': 'Aegiceras floridum',
  'avicennia_alba': 'Avicennia alba',
  'avicennia_marina': 'Avicennia marina',
  'avicennia_marina_var_rumphiana': 'Avicennia rumphiana',
  'avicennia_officinalis': 'Avicennia officinalis',
  'avicennia_rumphiana': 'Avicennia rumphiana',
  'bruguiera_cylindrica': 'Bruguiera cylindrica',
  'bruguiera_gymnorhiza': 'Bruguiera gymnorrhiza',
  'bruguiera_gymnorrhiza': 'Bruguiera gymnorrhiza',
  'bruguiera_sexangola': 'Bruguiera sexangula',
  'bruguiera_sexangula': 'Bruguiera sexangula',
  'camptostemon_philippinense': 'Camptostemon philippinensis',
  'camptostemon_phillipinensis': 'Camptostemon philippinensis',
  'camptostemon_philippinensis': 'Camptostemon philippinensis',
  'ceriops_tagal': 'Ceriops tagal',
  'ceriops_zippeliana': 'Ceriops zippeliana',
  'excoecaria_agallocha': 'Excoecaria agallocha',
  'heritiera_littoralis': 'Heritiera littoralis',
  'lumnitzera_littorea': 'Lumnitzera littorea',
  'lumnitzera_racemosa': 'Lumnitzera racemosa',
  'nypa_fruticans': 'Nypa fruticans',
  'osbornia_octodonta': 'Osbornia octodonta',
  'pemphis_acidula': 'Pemphis acidula',
  'rhizophora_apiculata': 'Rhizophora apiculata',
  'rhizophora_mucronata': 'Rhizophora mucronata',
  'rhizophora_stylosa': 'Rhizophora stylosa',
  'scyphiphora_hydrophyllacea': 'Scyphiphora hydrophylacea',
  'scyphiphora_hydrophylacea': 'Scyphiphora hydrophylacea',
  'sonneratia_alba': 'Sonneratia alba',
  'sonneratia_ovata': 'Sonneratia ovata',
  'xylocarpus_granatum': 'Xylocarpus granatum',
  'xylocarpus_moluccensis': 'Xylocarpus rumphii',
  'xylocarpus_rumphii': 'Xylocarpus rumphii',
};
